### PR TITLE
#3555 Add support for a routing chain configuration

### PR DESCRIPTION
**Pull request by keshy**
The change provides an implementation for a 'Map' semantics for model chaining. Similar to `Sequential` chains which can model fixed chain sequence implementation, the idea here is to enable consumers to be able to provide a map of chains and dynamically choose the right chain (or its variations) based on input. 

The decision of which chain to use will rely on a vector store lookup function that consumers can provide as a parameter to the chain definition. This will allow callers to use their own type of vector store and does not force a specific tech. 

To maintain the interface requirements of the chain, the chain attribution is added in the response text to indicate to the caller the downstream chain that's actually processing the request.

### CODE REVIEW

File: langchain/chains/llm_router_chain/base.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                m_name = last_chain_name[0]
+                _input = self.last_chain(_input)
+            else:
+                raise ValueError(
+                    "Suitable destination chain not found for this question. Distance computed from nearest match: " +
+                    str(distance))
+        self.callback_manager.on_text(
+            str(_input['text']), color=color_mapping[m_name], end="\n", verbose=self.verbose
+        )
+        return {
```
Comment (hwchase17) [inline langchain/chains/llm_router_chain/base.py:81]: lets respond with two keys - one for the chain name, one for the output

File: langchain/chains/llm_router_chain/base.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            raise ValueError("Vector lookup function not provided for this router chain.")
+        m_name, distance = self.vector_lookup_fn(query=[_input])
+        # picking a guardrail where if the LLM response is way off - then just use the same model as the previous
+        # one to continue conversing.
+        if self.chains.get(m_name) and distance <= self.fuzzy_match_threshold:
+            _input = self.chains[m_name](_input)
+        else:
+            if self.last_chain:
+                m_name = last_chain_name[0]
+                _input = self.last_chain(_input)
```
Comment (hwchase17) [inline langchain/chains/llm_router_chain/base.py:73]: should be `_output = `

File: langchain/chains/llm_router_chain/base.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    fuzzy_match_threshold = 1.5
+    """Default memory store."""
+    prompt: BasePromptTemplate = PROMPT
+    """Default conversation prompt to use."""
+    last_chain: Chain = None
+    chains: Dict[str, Chain]
+    strip_outputs: bool = False
+    input_key: str = "input"  #: :meta private:
+    output_key: str = "output"  #: :meta private:
+    vector_lookup_fn: FunctionType = None
```
Comment (hwchase17) [inline langchain/chains/llm_router_chain/base.py:30]: better typing here

if this is implemented using Vectorstores, can we use that?

File: langchain/chains/llm_router_chain/base.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+from langchain import LLMChain, BasePromptTemplate
+from langchain.chains.base import Chain
+from langchain.chains.conversation.prompt import PROMPT
+from langchain.input import get_color_mapping
+from langchain.memory import ConversationBufferWindowMemory
+from langchain.schema import BaseMemory
+
+
+class RouterChain(LLMChain):
```
Comment (hwchase17) [inline langchain/chains/llm_router_chain/base.py:14]: name is very general, should reflect more of the specifics (uses a vector store)

Comment (hwchase17) [inline langchain/chains/llm_router_chain/base.py:14]: should inherit from Chain instead

**Review by hwchase17 — COMMENTED** [review]
love the idea! may take a stab at this myself if i have time

File: langchain/chains/llm_router_chain/base.py (hunk 1 above)
Comment (keshy) [reply langchain/chains/llm_router_chain/base.py:81]: The issue is the extent of changes all the way down. I am still coming up to speed on this and the amount of changes got out of hand. Felt like I went the wrong way. It breaks the parent LLM class constraints for `predict` which expects only one key i.e. the output_key in the dict.

File: langchain/chains/llm_router_chain/base.py (hunk 3 above)
Comment (keshy) [reply langchain/chains/llm_router_chain/base.py:30]: I left this completely up to the caller. Did not want to enforce a vector store (although my initial change did have that). I feel for some cases even keeping those in memory is a possibility which does not require a heavy vector store integration.

File: langchain/chains/llm_router_chain/base.py (hunk 4 above)
Comment (keshy) [reply langchain/chains/llm_router_chain/base.py:14]: ok - maybe that'd unlock the issue with the output constraint.

File: langchain/chains/llm_router_chain/base.py (hunk 2 above)
Comment (keshy) [reply langchain/chains/llm_router_chain/base.py]: Done

File: langchain/chains/llm_router_chain/base.py (hunk 4 above)
Comment (keshy) [reply langchain/chains/llm_router_chain/base.py]: Took care of it - thanks. This helped me figure out what's required to maintain the `lastChain` context through history.

**Comment by keshy** [comment]
Can I get a review please? @dev2049

**Review by hwchase17 — COMMENTED** [review]
we've recently added a `RouterChain`.

the correct way to do this would be to inherit from that chain, and add the main logic from here to there. the main logic here seems to be using embeddings aka retrieval for choosing the routing logic

i took a stab at implementing tht here: https://github.com/hwchase17/langchain/pull/4644

**Comment by keshy** [comment]
Ok - am confused what I should do with my PR. Just close it? It Seems like
you implemented the same thing I was trying to implement so not sure why
I'd extend it from the new chain there if that does this same thing.

Will close this PR.

**Comment by dev2049** [comment]
closing per above ^

**Pull request closed without merging**